<a href="https://colab.research.google.com/github/Elena-Shabanova/-Python/blob/Intenciv_MFTI/%D0%94%D0%97_%22%5Bhomework%5Dpandas_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<p style="align: center;"><img src="https://static.tildacdn.com/tild6636-3531-4239-b465-376364646465/Deep_Learning_School.png", width=600, height=300></p>

<h3 style="text-align: center;"><b>Школа глубокого обучения ФПМИ МФТИ</b></h3>

<h2 style="text-align: center;"><b>Домашнее задание. Библиотека <a href="http://pandas.pydata.org/">pandas</a></b></h2>

В этом задании вы проанализируете данные из Appstore

<img src="https://drive.google.com/uc?export=view&id=1F1Qajgj-dC4o4VI3Wm7fIGlOJA0a60_B" width=300 height=300>


In [51]:
#!pip install pandas

In [52]:
import pandas as pd
import numpy as np

### Данные для задания: https://drive.google.com/file/d/1JY8l5nSu9O4GtMDpaOfH2Oowlpza3U-c/view?usp=sharing

In [70]:
data = pd.read_csv('https://drive.google.com/uc?id=1JY8l5nSu9O4GtMDpaOfH2Oowlpza3U-c')

Посмотрим, что же такое переменная `data`:

In [61]:
data

User_ID Product_ID  ... Product_Category_3 Purchase
0       1000001  P00069042  ...                NaN     8370
1       1000001  P00248942  ...               14.0    15200
2       1000001  P00087842  ...                NaN     1422
3       1000001  P00085442  ...                NaN     1057
4       1000002  P00285442  ...                NaN     7969
...         ...        ...  ...                ...      ...
537572  1004737  P00193542  ...                NaN    11664
537573  1004737  P00111142  ...               16.0    19196
537574  1004737  P00345942  ...                NaN     8043
537575  1004737  P00285842  ...                NaN     7172
537576  1004737  P00118242  ...                NaN     6875

[537577 rows x 12 columns]

In [36]:
type(data)

pandas.core.frame.DataFrame

---

#### Вопрос

О чём данные? (Hint: https://www.kaggle.com/sdolezel/black-friday)

База данных о покупках 3-х категорий товаров жителей городов/регионов (537576 записи) с разбивкой по полу, возрасной категории, профессии, семейному положению и т.д. (по 12 критериям/категориям)

---

#### Основное задание (в тесте надо будет вставлять ответы для этих пунктов):

*Примечание:* не бойтесь гуглить и заглядывать в "Полезные ссылки" для того, чтобы выполнить какие-то задания. Возможно, на семинаре не было какого-то нужного метода, но он находится в поисковике за 2 минуты.

**0).** Сколько всего возрастных категорий?  

In [90]:
data['Age'].value_counts()

26-35    214690
36-45    107499
18-25     97634
46-50     44526
51-55     37618
55+       20903
0-17      14707
Name: Age, dtype: int64

**1).** Сколько строк с мужчинами из категории города A? (речь не об уникальных ID мужчин, а о количестве строк)  

In [83]:
#data.groupby(['City_Category'])['Gender'].sum() # Суммирует, но по всем городам
#len(data[(data['City_Category']=='A') & (data['Gender'] == 'M')]) # Не пошло
data['Sex'] = (data['Gender'] == 'M').astype(int)
len(data[(data['City_Category']=='A') & (data['Gender'] == 1)])

109831

**2).** Сколько женщин от 46 до 50, потративших (столбец Purchase) больше 20000 (условных единиц, в данном случае)?   (речь не об уникальных ID, а о количестве строк)  

In [87]:
len(data[(data['Age']=='46-50') & (data['Gender'] == 0) & (data['Purchase']>20000)])

270

**3).** Сколько NaN'ов в столбце Product_Category_3?  

In [92]:
#data['Product_Category_3'].isna().sum() # Для проверки. Тут два способа решений
len(data) - data['Product_Category_3'].count() 

373299

**4).** Какую долю (вещественное число от 0 до 1, округлить до 4-го знака) от всех покупателей составляют ВМЕСТЕ мужчины от 26 до 35 лет и женщины старше 36 лет (то есть нужно учесть несколько возрастных категорий)? (речь не об уникальных ID, а о количестве таких строк)  

In [98]:
val1=data[(data['Age'] == '26-35') & (data['Gender'] == 1)]
val2=data[(data['Age'] == '46-54') & (data['Age'] == '36-45') & (data['Age'] == '55+') & (data['Gender'] == 0)]
print(round((val1.shape[0]+val2.shape[0])/data.shape[0], 4))

0.3076


---

Больше про pandas можно найти по этом полезным ссылкам:

* Официальные туториалы: http://pandas.pydata.org/pandas-docs/stable/tutorials.html

* Статья на Хабре от [OpenDataScience сообщества](http://ods.ai/)**:** https://habr.com/company/ods/blog/322626/

* Подробный гайд: https://media.readthedocs.org/pdf/pandasguide/latest/pandasguide.pdf

Главное в работе с новыми библиотеками -- не бояться тыкать в разные функции, смотреть типы возвращаемых объектов и активно пользоваться Яндексом, а ещё лучше понимать всё из docstring'а (`Shift+Tab` при нахождении курсора внутри скобок функции).